### PR TITLE
[WIP] Add Import Service and IngestSST/UploadSST RPC

### PR DESCRIPTION
**Pull request by huachaohuang**
Allow a client to upload SST files to TiKV, and then ingest them to RocksDB.

This PR follows https://github.com/pingcap/tikv/pull/2420, and will be rebased later.
Put it here for now for easy review.

### CODE REVIEW

File: src/raftstore/store/worker/apply.rs [hunk 1]
```diff
@@ -1176,6 +1187,57 @@ impl ApplyDelegate {
...
+
+    fn handle_ingest_sst(&mut self, req: &Request) -> Result<Response> {
+        let mut cfname = None;
+        let ingest = req.get_ingest_sst();
+        let region_id = self.region.get_id();
+        let region_epoch = self.region.get_region_epoch();
+
+        let mut ingest_paths = vec![];
+        for h in ingest.get_handles() {
+            if h.get_region_id() != region_id ||
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: can we meet mismatched region ID?

File: src/raftstore/store/worker/apply.rs [hunk 2]
```diff
@@ -1176,6 +1187,57 @@ impl ApplyDelegate {
...
+                h.get_region_epoch().get_conf_ver() != region_epoch.get_conf_ver() ||
+                h.get_region_epoch().get_version() != region_epoch.get_version()
+            {
+                let error = "ingest sst files".to_owned();
+                let new_regions = vec![self.region.clone()];
+                return Err(Error::StaleEpoch(error, new_regions));
+            }
+            match cfname {
+                Some(cf) => if h.get_cfname() != cf {
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: I prefer get_cf_name

File: src/raftstore/store/worker/apply.rs [hunk 3]
```diff
@@ -1176,6 +1187,57 @@ impl ApplyDelegate {
 
         Ok(resp)
     }
+
+    fn handle_ingest_sst(&mut self, req: &Request) -> Result<Response> {
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: we may meet the corner case that after handle_ingest_sst but fail to save the apply index, we will retry applying the command again, how do we do?

File: src/raftstore/store/worker/apply.rs (hunk 1 above)
Comment (huachaohuang) [reply src/raftstore/store/worker/apply.rs]: Yep, with a faulted client.

File: src/raftstore/store/worker/apply.rs (hunk 3 above)
Comment (huachaohuang) [reply src/raftstore/store/worker/apply.rs]: Seems we will not retry, we just panic if failed to write to the engine.

Comment (huachaohuang) [reply src/raftstore/store/worker/apply.rs]: I got it, it will still retry on restart after panicked.

Comment (huachaohuang) [reply src/raftstore/store/worker/apply.rs]: I think we can create hard links before ingesting, but we need to clean up them after they are applied successfully.

Comment (siddontang) [reply src/raftstore/store/worker/apply.rs]: This is the same problem as snapshot applying. 

@zhangjinpeng1987 we should support updating sequence number ASAP.

**Comment by huachaohuang** [comment]
Moved to another place now.

**Pull request closed without merging**